File: puzzles/ts/skynet-revolution1/skynet-revolution1.ts
function sever(n1: number, n2: number): void {
  console.log(`${n1} ${n2}`);
}

function deleteLink(graph: object, n1: number, n2: number): void {
  graph[n1].splice(graph[n1].indexOf(n2), 1);
  graph[n2].splice(graph[n2].indexOf(n1), 1);
}

function blockDirectLink(graph: object, gateways: number[], agentNodeId: string): boolean {
  // check if the current agent node is linked to a gateway
  let gateway: number = -1;
  for (let neighbor of graph[agentNodeId]) {
    if (gateways.indexOf(neighbor) != -1) {
      sever(Number(agentNodeId), neighbor);
      gateway = neighbor;
    }
  }
  if (gateway !== -1) {
    deleteLink(graph, Number(agentNodeId), gateway);
  }
  return gateway !== -1;
}

function blockRandomLink(graph: object, gateways: number[]): void {
  for (let id in graph) {
    if (graph[id].length > 0) {
      sever(Number(id), graph[id][0]);
      deleteLink(graph, Number(id), graph[id][0]);
      return;
    }
  }
}

const inputs: string = readline().split(' ');
const nbNodes = Number(inputs[0]); // the total number of nodes in the level, including the gateways
const nbLinks = Number(inputs[1]); // the number of links
const nbGateways = Number(inputs[2]); // the number of exit gateways

let graph: object = {}; // map node id => node links set
let gateways: number[] = [];

for (let i: number = 0; i < nbLinks; i++) {
  const inputs: string = readline().split(' ');
  const n1 = Number(inputs[0]); // n1 and n2 defines a link between these nodes
  const n2 = Number(inputs[1]);
  
  if (! (n1 in graph)) {
    graph[n1] = []
  }
  graph[n1].push(n2);

  if (! (n2 in graph)) {
    graph[n2] = []
  }
  graph[n2].push(n1);
}

for (let i: number = 0; i < nbGateways; i++) {
  const gateway = Number(readline()); // the id of a gateway node
  gateways.push(gateway);
}

// game loop
while (true) {
  const agentNodeId:string = readline();
  if (!blockDirectLink(graph, gateways, agentNodeId)) {
    blockRandomLink(graph, gateways);
  };
}

File: puzzles/python3/skynet-revolution1/skynet_revolution1.py
from collections import deque


def bfs(graph, gateways, agent_node_id):
    visited = set()
    queue = deque()
    came_from = {}  # node_id => parent node id on the shortest path
    queue.append(agent_node_id)
    visited.add(agent_node_id)

    while len(queue) != 0:
        node_id = queue.popleft()
        for neighbor_id in graph[node_id]:
            if neighbor_id not in visited:
                visited.add(neighbor_id)
                queue.append(neighbor_id)
                came_from[neighbor_id] = node_id
                if neighbor_id in gateways:
                    return (came_from, neighbor_id)


def reconstruct_path(graph, came_from, neighbor_id):
    current_id = neighbor_id
    stack = []

    while current_id in came_from:
        stack.append(current_id)
        current_id = came_from[current_id]
    stack.append(current_id)
    return stack


# read game input
graph = {}  # node_id => links set
gateways = set()
nb_nodes, nb_links, nb_gateways = map(int, input().split())
for i in range(nb_links):
    # n1 and n2 defines a link between these nodes
    n1, n2 = map(int, input().split())
    if n1 not in graph:
        graph[n1] = set()
    if n2 not in graph:
        graph[n2] = set()
    graph[n1].add(n2)
    graph[n2].add(n1)
for i in range(nb_gateways):
    gateways.add(int(input()))

# game loop
while True:
    agent_node_id = int(input())  # node id on which the Skynet agent is located
    came_from, neighbor_id = bfs(graph, gateways, agent_node_id)
    path = reconstruct_path(graph, came_from, neighbor_id)
    second_node_id = path[-2]
    graph[agent_node_id].remove(second_node_id)
    graph[second_node_id].remove(agent_node_id)
    # the indices of the nodes you wish to sever the link between
    print("{} {}".format(agent_node_id, second_node_id))
